src/overlays/actors/ovl_En_Wf/z_en_wf.c

```diff
@@ -137,81 +144,1338 @@ static ColliderCylinderInit D_80B37A5C = {
+}
+
+void EnWf_Init(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    EnWf* this = THIS;
+
+    Actor_ProcessInitChain(thisx, sInitChain);
+    thisx->colChkInfo.damageTable = &sDamageTable;
+    ActorShape_Init(&thisx->shape, 0.0f, ActorShadow_DrawCircle, 0.0f);
+    thisx->focus.pos = thisx->world.pos;
+    thisx->colChkInfo.mass = MASS_HEAVY;
+    thisx->colChkInfo.health = 8;
+    thisx->colChkInfo.cylRadius = 50;
+    thisx->colChkInfo.cylHeight = 100;
+    this->switchFlag = (thisx->params >> 8) & 0xFF;
+    thisx->params &= 0xFF;
+    this->eyeIndex = 0;
+    this->unk_2F4 = 10.0f; // Set and not used
+
+    Collider_InitJntSph(globalCtx, &this->colliderSphere);
+    Collider_SetJntSph(globalCtx, &this->colliderSphere, thisx, &sJntSphInit, this->colliderSphereElements);
+    Collider_InitCylinder(globalCtx, &this->colliderCylinder1);
+    Collider_SetCylinder(globalCtx, &this->colliderCylinder1, thisx, &sCylinderInit1);
+    Collider_InitCylinder(globalCtx, &this->colliderCylinder2);
+    Collider_SetCylinder(globalCtx, &this->colliderCylinder2, thisx, &sCylinderInit2);
+
+    if (thisx->params == WOLFOS_NORMAL) {
+        SkelAnime_InitFlex(globalCtx, &this->skelAnime, &gWolfosNormalSkel, &gWolfosWaitingAnim, this->jointTable,
+                           this->morphTable, WOLFOS_LIMB_MAX);
+        Actor_SetScale(thisx, 0.0075f);
+        thisx->naviEnemyId = 0x4C; // Wolfos
+    } else {                       // WOLFOS_WHITE
+        SkelAnime_InitFlex(globalCtx, &this->skelAnime, &gWolfosWhiteSkel, &gWolfosWaitingAnim, this->jointTable,
+                           this->morphTable, WOLFOS_LIMB_MAX);
+        Actor_SetScale(thisx, 0.01f);
+        this->colliderSphere.elements[0].info.toucher.damage = this->colliderSphere.elements[1].info.toucher.damage = 8;
+        thisx->naviEnemyId = 0x57; // While Wolfos
+    }
+
+    EnWf_SetupWaitToAppear(this);
+
+    if ((this->switchFlag != 0xFF) && Flags_GetSwitch(globalCtx, this->switchFlag)) {
+        Actor_Kill(thisx);
+    }
+}
+
+void EnWf_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    EnWf* this = THIS;
+
+    Collider_DestroyJntSph(globalCtx, &this->colliderSphere);
+    Collider_DestroyCylinder(globalCtx, &this->colliderCylinder1);
+    Collider_DestroyCylinder(globalCtx, &this->colliderCylinder2);
+
+    if ((this->actor.params != 0) && (this->switchFlag != 0xFF)) {
+        func_800F5B58();
+    }
+
+    if (this->actor.parent != NULL) {
+        EnEncount1* parent = (EnEncount1*)this->actor.parent;
+
+        if (parent->actor.update != NULL) {
+            if (parent->curNumSpawn > 0) {
+                parent->curNumSpawn--;
+            }
+
+            osSyncPrintf("\n\n");
+            // "☆☆☆☆☆ Number of concurrent events ☆☆☆☆☆"
+            osSyncPrintf(VT_FGCOL(GREEN) "☆☆☆☆☆ 同時発生数 ☆☆☆☆☆%d\n" VT_RST, parent->curNumSpawn);
+            osSyncPrintf("\n\n");
+        }
+    }
+}
+
+s32 EnWf_ChangeAction(GlobalContext* globalCtx, EnWf* this, s16 arg2) {
```

Comment:
I think `Change` is better because it returns `true` if it does choose something, not sure that's so obvious with `Choose`.

I don't think `mustChoose` can be right though: all it does is widen the checks it considers before returning `false`.

Reply:
about Change/Choose whatever I guess

about `mustChoose` yeah exactly, notice that if it is true then the function will always pick an action and return true.
I think this makes sense since `EnWf_ChangeAction(globalCtx, this, false)` is usually used at the start of actions (some kind of "emergency react" thing),
while `EnWf_ChangeAction(globalCtx, this, true);` is often used (eg in `EnWf_Stunned`) to inconditionally pick a new action

Reply:
```c
    if (arg2) {
        s16 playerFacingAngleDiff;
        if (playerYawDiff >= 0x1B58) {
            EnWf_SetupSidestep(this, globalCtx);
            return true;
        }

        playerFacingAngleDiff = player->actor.shape.rot.y - this->actor.shape.rot.y;

        if (((this->actor.xzDistToPlayer <= 80.0f) && !Actor_OtherIsTargeted(globalCtx, &this->actor)) &&
            (((globalCtx->gameplayFrames % 8) != 0) || (ABS(playerFacingAngleDiff) < 0x38E0))) {
            EnWf_SetupSlash(this);
            return true;
        } else {
            EnWf_SetupRunAroundPlayer(this);
            return true;
        }
    }```
It seems to me that it can definitely escape from this conditional without choosing an action.

Reply:
Can you highlight a situation where control would exit this if? I can't see it

let me label the blocks so we can talk about this easier
```c
    if (arg2) { // "main if"
        s16 playerFacingAngleDiff;
        if (playerYawDiff >= 0x1B58) { // "first if"
            EnWf_SetupSidestep(this, globalCtx);
            return true;
        }

        playerFacingAngleDiff = player->actor.shape.rot.y - this->actor.shape.rot.y;

        // "second if"
        if (((this->actor.xzDistToPlayer <= 80.0f) && !Actor_OtherIsTargeted(globalCtx, &this->actor)) &&
            (((globalCtx->gameplayFrames % 8) != 0) || (ABS(playerFacingAngleDiff) < 0x38E0))) {
            EnWf_SetupSlash(this);
            return true;
        } else { // "final else"
            EnWf_SetupRunAroundPlayer(this);
            return true;
        }
    }
```

So what I'm seeing is that if control enters the "main if", it can either enter the "first if" (and choose an action `EnWf_SetupSidestep`, return true) or not.
If it does not enter the "first if" then it reaches the "second if", which it may enter and choose the action `EnWf_SetupSlash`, or it will enter the "final else" and also choose an action `EnWf_SetupRunAroundPlayer`

Reply:
Oh, you're quite right, I can't read 😓 . I'll annotate it so other people don't miss it.

Reply:
Came up with a better way: just not use an else in the first place.